include/real/real_math.hpp

```diff
@@ -0,0 +1,237 @@
+		 * @author: Vikram Singh Chundawat
+		 **/
+		template<typename T>
+		inline exact_number<T> tangent(exact_number<T> x, size_t max_error_exponent, bool upper){
+			exact_number<T> result = sine(x, max_error_exponent, upper);
+			exact_number<T> cos = cosine(x,  max_error_exponent, upper);
+			result.divide_vector(cos, max_error_exponent, upper);
+			result = result.up_to(max_error_exponent, upper);
+			return result; 
+		}
+
+		/**
+		 *  COTANGENT FUNCTION USING TAYLOR EXPANSION
+		 * @brief: calculates cot(x) of a exact_number using taylor expansion
+		 * @param: x: the exact_number.
+		 * @param: max_error_exponent: Absolute Error in the result should be < 1*base^(-max_error_exponent)
+		 * @param:  upper: if true: error lies in [0, +epsilon]
+		 *                  else: error lies in [-epsilon, 0], here epsilon = 1*base^(-max_error_exponent)
+		 * @author: Vikram Singh Chundawat
+		 **/
+		template<typename T>
+		inline exact_number<T> cotangent(exact_number<T> x, size_t max_error_exponent, bool upper){
+			exact_number<T> result = cosine(x, max_error_exponent, upper);
+			exact_number<T> sin = sine(x,  max_error_exponent, upper);
+			result.divide_vector(sin, max_error_exponent, upper);
+			result = result.up_to(max_error_exponent, upper);
+			return result; 
+		}
+
+		/**
+		 *  SECANT FUNCTION USING TAYLOR EXPANSION
+		 * @brief: calculates sec(x) of a exact_number using taylor expansion
+		 * @param: x: the exact_number.
+		 * @param: max_error_exponent: Absolute Error in the result should be < 1*base^(-max_error_exponent)
+		 * @param:  upper: if true: error lies in [0, +epsilon]
+		 *                  else: error lies in [-epsilon, 0], here epsilon = 1*base^(-max_error_exponent)
+		 * @author: Vikram Singh Chundawat
+		 **/
+		template<typename T>
+		inline exact_number<T> secant(exact_number<T> x, size_t max_error_exponent, bool upper){
+			exact_number<T> result("1");
+			exact_number<T> cos = cosine(x, max_error_exponent, upper);
+			result.divide_vector(cos, max_error_exponent, upper);
+			result = result.up_to(max_error_exponent, upper);
+			return result;
+		}
+
+		/**
+		 *  COSECANT FUNCTION USING TAYLOR EXPANSION
+		 * @brief: calculates cosec(x) of a exact_number using taylor expansion
+		 * @param: x: the exact_number.
+		 * @param: max_error_exponent: Absolute Error in the result should be < 1*base^(-max_error_exponent)
+		 * @param:  upper: if true: error lies in [0, +epsilon]
+		 *                  else: error lies in [-epsilon, 0], here epsilon = 1*base^(-max_error_exponent)
+		 * @author: Vikram Singh Chundawat
+		 **/
+		template<typename T>
+		inline exact_number<T> cosecant(exact_number<T> x, size_t max_error_exponent, bool upper){
+			exact_number<T> result("1");
+			exact_number<T> sin = sine(x, max_error_exponent, upper);
+			result.divide_vector(sin, max_error_exponent, upper);
+			result = result.up_to(max_error_exponent, upper);
+			return result;
+		}

```

Comment:
same as previous comments

Reply:
Again, Calculating it by reciprocating sin(x) is easier and efficient.